### ISSUE TITLE
WASM: Support for Wasi 0.2.0 on wasmtime17

### ISSUE THREAD

**Issue by tippexs** [nginx/unit#1098]
Aa user I want to be able to load a WebAssembly Component build with the wasi 0.2.0 snapshots on Unit. 

We don't have to support previous snapshots like `0.2.0-rc-2023-10-18` or `0.2.0-rc-2023-11-10`

## Technical Requirements

For the next release it will be required to:
- [ ] support for wasi 0.2.0 based WebAssembly components (wasmtime 17) in Units `wasm-wasi-http` language module
- [ ] be able to access files on disk from inside the WebAssembly Component.
- [ ] Access to other resources like HTTP or TCP endpoints from inside the Module can be implemented afterwards. No strong need for the next release 1.32.
- [ ] Naming: `wasm-wasi-component` - [Read more](https://github.com/nginx/unit/issues/1098#issuecomment-1930901281)
- [ ] the loaded WebAssembly Component should hook into the `restart` handlers. This means if the WebAssembly module changes on disk, a `restart` of the Unit `application` should pick up the new WebAssembly Component Binary from disk. However for developers it will be important that the WebAssembly component will automatically picked up once changed. We should challenge this.

## Documentation

- [x] Update our documentation and examples match the WASI-Preview 0.2.0 Snapshot

## Test Case

- [ ] Deploy an echo-server WebAssembly component and return Request meta data like `URI`, `Protocol Version`, `Host`, `Headers`
- [ ] Deploy an Test-Component that will be able to receive Files and store them on Disk. File size should vary from 1MB - <= 600MB. We had some technical issues in the past handling "big" files. This must be tested.

## Misc

- [x] An example implementation of a wasi-http component can be found [here](https://github.com/tippexs/hello-wasi-http). This is not an echo server but will work for a first test / getting started point.

**Comment by callahad** [comment]
Context:
- The WASI subgroup voted to launch WASI Preview2 (0.2.0) last Thursday, 25 January.
- [Wasmtime v17.0.0 was released](https://github.com/bytecodealliance/wasmtime/releases/tag/v17.0.0) that same day with the Component Model, `wasi:cli`, and `wasi:http` worlds on by default.
    - ...though embeddings still need to opt-in to the component model
- Direct support for `wasm32-wasi-preview2` should land in Rust nightlies soon: https://github.com/rust-lang/rust/pull/119616

We need to update Unit's wasm-wasi module to Wasmtime v17, and then verify/update our docs and examples.

**Comment by ac000** [comment]
Language module is updated to wasmtime 17 and can load and run a WASI 0.2.0 component.

**Comment by tippexs** [comment]
After a chat with @ospencer the new WebAssembly Module for Unit should be labeled as `wasm32-wasi-component`

By using this an user will be able to see just from the name what this language module will be able to run on Unit. A WebAssembly 32bit Wasi Component. 

@ospencer  Thanks for the valuable input on this!

**Comment by ospencer** [comment]
If it weren't for the verbosity, I'd also consider `wasm32-wasi-preview2-component` to make it exceptionally clear what it's able to run.

**Comment by ac000** [comment]
On Tue, 06 Feb 2024 14:50:02 -0800
Timo Stark ***@***.***> wrote:

> After a chat with @ospencer the new WebAssembly Module for Unit should be labeled as `wasm32-wasi-component`

So this should be the module 'type' that you put in the Unit config?

Also where it should live in the source tree,
'src/wasm32-wasi-component'?

> By using this an user will be able to see just from the name what this language module will be able to run on Unit. A WebAssembly 32bit Wasi Component.
> 
> @ospencer  Thanks for the valuable input on this!

What about if wasm64 becomes a thing? Would that require a whole new
language module or would the existing module be able to support that?
In which case a slightly more generic name might be more appropriate.

**Comment by ospencer** [comment]
Wasm64 will indeed be a thing. As the proposal stands, it _should_ be able to be the same module. Up to you all how you want to organize the modules, though.

**Comment by tippexs** [comment]
IF we name it `wasm-wasi-component` how do we tell the user what to use? `32` or `64`? Does the user have to configure it actively or is this something we can find out by looking into the presented component using `wit-tools` in the language module? If that is the case I am fine by using `wasm-wasi-component` with the configuration we are currently using in our examples.

@ac000 I think `src/wasm-wasi-component` is a good place to put the language module but I will leave this to you on how to organize the source tree. My only concern really was / is about exactly this

> So this should be the module 'type' that you put in the Unit config?
Yes. We must be very clear whats actually being used to load the component. I do like the `component` configuration option btw.

```json
  "applications": {
          "wasm": {
              "type": "wasm-wasi-component",
              "component": "/path/to/power-spin-rs/application/application.wasm"
          }
      }
```

**Comment by ac000** [comment]
(You beat me to it!)...

The thing is, currently a language module can only register itself as a single 'type'. So if this module registers itself as `wasm32-wasi-component` then if we want to also support `wasm64-wasi-component` as a module type, you know because people won't intuitively think that wasm32... will support wasm64..., that would require a whole separate module.

Given that then perhaps just `wasm-wasi-component` is more future proof...

**Comment by ospencer** [comment]
`wasm-wasi-component` works fine.

**Comment by tippexs** [comment]
Was released with 1.32.0!
